File: homework3/steven_law/q2_Heap.cpp
// Time spent: 36 mins 22 seconds

#include <iostream>
#include <vector>
using namespace std;

class Heap {
private:
    vector<int> arr;

    // bubble up at index i
    void bubbleUp(int i){
        while(i > 0){
            int parent = (i - 1) / 2;
            if(arr[i] < arr[parent]){
                swap(arr[i], arr[parent]);
                i = parent;
            } else {
                break;
            }
        }
    }

    // bubble down from index i
    void bubbleDown(int i){
        int n = arr.size();
        while(true){
            int left = 2*i + 1;
            int right = 2*i + 2;
            int smallest = i;

            if(left < n && arr[left] < arr[smallest]) smallest = left;
            if(right < n && arr[right] < arr[smallest]) smallest = right;

            if(smallest == i) break;
            swap(arr[i], arr[smallest]);
            i = smallest;
        }
    }

public:
    // returns the min (top) element in the heap
    int top(){
        if(arr.empty()) return -1;
        return arr[0];
    }

    // adds int x to the heap in the appropriate position
    void insert(int x){
        arr.push_back(x);
        bubbleUp(arr.size() - 1);
    }

    // removes the min (top) element in the heap
    void remove(){
        if(arr.empty()) return;
        arr[0] = arr.back();
        arr.pop_back();
        if(!arr.empty()) bubbleDown(0);
    }
};

int main(){
    Heap h;
    h.insert(5);
    h.insert(3);
    h.insert(8);
    h.insert(1);
    h.insert(9);
    h.insert(2);

    cout << "Top: " << h.top() << " (Expected: 1)" << endl;
    h.remove();
    cout << "Top after remove: " << h.top() << " (Expected: 2)" << endl;
    h.remove();
    cout << "Top after remove: " << h.top() << " (Expected: 3)" << endl;
    h.insert(0);
    cout << "Top after insert 0: " << h.top() << " (Expected: 0)" << endl;

    return 0;
}

File: homework3/danny_tat/q2_heap.py
class Heap:
    def __init__(self):
        self.arr = []

    def top(self):
        # Returns the min (top) element
        return self.arr[0] if self.arr else None

    def insert(self, x):
        # Adds x to the heap in the correct position
        self.arr.append(x)
        self._sift_up(len(self.arr) - 1)

    def remove(self):
        # Removes the min (top) element
        if not self.arr:
            return
        # Swap root with last element, then sift down
        self.arr[0], self.arr[-1] = self.arr[-1], self.arr[0]
        self.arr.pop()
        if self.arr:
            self._sift_down(0)

    def _sift_up(self, i):
        while i > 0:
            parent = (i - 1) // 2
            if self.arr[i] < self.arr[parent]:
                self.arr[i], self.arr[parent] = self.arr[parent], self.arr[i]
                i = parent
            else:
                break

    def _sift_down(self, i):
        n = len(self.arr)
        while True:
            smallest = i
            left, right = 2 * i + 1, 2 * i + 2
            if left < n and self.arr[left] < self.arr[smallest]:
                smallest = left
            if right < n and self.arr[right] < self.arr[smallest]:
                smallest = right
            if smallest == i:
                break
            self.arr[i], self.arr[smallest] = self.arr[smallest], self.arr[i]
            i = smallest

# Time: 40 min
